### ISSUE TITLE
What's the correct way to break the PipeReader loop?

### ISSUE THREAD

**Issue by GF-Huang** [dotnet/docs#22833]
This [page](https://docs.microsoft.com/en-us/dotnet/standard/io/pipelines#pipereader) show many incorrect demonstrations.

But what's the correct way?

Is this the only correct way?

![image](https://user-images.githubusercontent.com/4510984/107858778-e0b0ae80-6e70-11eb-8d0a-2251829337b2.png)

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 2f5384a7-5e96-d977-7748-317a2c758f63
* Version Independent ID: 06e72212-a052-a351-0e0b-1b002ce214de
* Content: [I/O pipelines - .NET](https://docs.microsoft.com/en-us/dotnet/standard/io/pipelines)
* Content Source: [docs/standard/io/pipelines.md](https://github.com/dotnet/docs/blob/master/docs/standard/io/pipelines.md)
* Product: **dotnet-fundamentals**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

**Comment by davidfowl** [comment]
Yes assuming you want to throw for incomplete messages.

**Comment by GF-Huang** [comment]
> Yes assuming you want to throw for incomplete messages.

So, `result.IsComplete && buffer.Length == 0` means all messages are read complete?

**Comment by davidfowl** [comment]
Yes but look at the examples. There are 2 code samples. One for reading a single message and one for reading many messages. The code is different depending on what you're trying to do.

**Comment by GF-Huang** [comment]
> Yes but look at the examples. There are 2 code samples. One for reading a single message and one for reading many messages. The code is different depending on what you're trying to do.

I have another question is that why the single message example never call `reader.Complete`? Will this make memory leak?

**Comment by davidfowl** [comment]
When reading a single message at a time you don't want to complete the reader, that would mean you can't read anymore messages. In the case of reading a single message, the caller is responsible for completing the reader.

**Comment by GF-Huang** [comment]
Got it, thanks your guide.

**Comment by Rick-Anderson** [comment]
Opening as we can probably add some of this to the doc.

**Comment by calrsom** [comment]
The primary way to break a PipeReader loop is to return one or more messages.  The loop is there to keep reading from the pipe until you can parse one.  The method is designed to block until it can return messages.

The `IsCompleted` check is there to handle the lesser case that the Stream is finished (for something like a file) or closed/disconnected (for something like a network connection).
